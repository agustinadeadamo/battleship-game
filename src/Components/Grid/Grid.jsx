/**
 * @desc Dependencies
 */
import React from 'react';
import PropTypes from 'prop-types';

/**
 * @desc Styles
 */
import {
  GridComponent,
} from './style';

const Grid = (props) => {
  // Destructuring props
  const {
    gridName,
    onClickCell,
    enableGrid,
    grid,
  } = props;

  return (
    <div>
      <GridComponent data-testid="grid-component" gridName={gridName} className={`grid-${gridName}`}>
        <div className="content">
        {
          grid && grid.map((cell) => (
            <div
              key={cell.id}
              role="presentation"
              onClick={() => enableGrid && onClickCell(cell)}
              id={cell.id}
              className={`cell ${cell.status}`}
              data-testid="cell-component"
            />
          ))
        }
        </div>
      </GridComponent>
    </div>
  );
};

Grid.propTypes = {
  gridName: PropTypes.string.isRequired,
  onClickCell: PropTypes.func.isRequired,
  enableGrid: PropTypes.bool,
  grid: PropTypes.arrayOf(PropTypes.object).isRequired,
};

Grid.defaultProps = {
  enableGrid: false,
};

export default Grid;
