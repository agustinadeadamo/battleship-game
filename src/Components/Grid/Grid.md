# Grid Component
Grid that has 100 cells, 10 x 10, and shows where the boats are positioned, if the boat was hit, destroyed or if there is sea

## Props

+ **grid** Should get an array throught the prop "grid" with ten object like the following one:
{
  id: 1,
  status: 'empty',
  taken: false,
  cellContent: 'sea',
  content: 'cruisers-1',
}
+ **status** should be 'empty', 'hit', 'destroyed' or 'missed', it indicate the status and gives style
+ **taken** indicates if there is a ship in that cell
+ **cellContent** indicates the kind of content in that cell, for example 'cruisers-1-1' that it's the cell 1 of the cruicer 1
+ **taken** indicates if there is a ship in that cell