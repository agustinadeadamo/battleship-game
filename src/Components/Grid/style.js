/**
 * @desc Dependencies
 */
import styled from 'styled-components';

/**
 * @desc Variabled
 */
import Colors from '../../MainStyles/Variables';

export const GridComponent = styled.div`
  width: 100%;
  display: inline-block;
  position: relative;
  display: flex;
  flex-wrap: wrap;

  &:before {
    content: "";
    display: block;
    padding-top: 100%;
  }

  .content {
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    display: flex;
    flex-wrap: wrap;
  }
  
  &.grid-user {
    .hit {
      background-color: ${Colors.orange};
    }

    .missed {
      background-color: ${Colors.lightBlue};
    }

    .destroyed {
      background-color: ${Colors.red};
    }

    .taken {
      background-color: ${Colors.grey};
    }
  }

  .cell {
    width: 10%;
    height: 10%;
    ${(props) => props.enableGrid && 'cursor: pointer;'}

    &.empty {
      border: 2px solid ${Colors.grey};
    }

  }
`;

export default GridComponent;
