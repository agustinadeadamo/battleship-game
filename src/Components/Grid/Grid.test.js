/**
 * @desc Dependencies
 */
import { create, act } from 'react-test-renderer';
import React from 'react';
import { render, fireEvent } from '@testing-library/react';

/**
 * @desc Components
 */
import Grid from './Grid';

let grid = [{
  id: 1,
  status: 'empty',
  taken: false,
  cellContent: 'sea',
}]

describe('[Grid Component]',() => {

  it('Renders correctly', () => {
    const {queryByTestId} = render(<Grid gridName="user"  grid={grid} onClickCell={() => {}}/>);
    expect(queryByTestId('grid-component')).toBeTruthy();
  });
  
  it('Matches snapshot', () => {
    const input = create(<Grid gridName="user"  grid={grid} onClickCell={() => {}} />).toJSON;
    expect(input).toMatchSnapshot();
  });

  it("Should call prop function onClickCell", () => {
    const onClickCell = jest.fn();
    const { getByTestId } = render(<Grid gridName="user" grid={grid} enableGrid={true}  onClickCell={onClickCell} />);
    // Gets cell item
    const cell = getByTestId("cell-component");

    // Fires onClick event in cell item
    act(() => {
      fireEvent.click(cell);
    });

    expect(onClickCell).toHaveBeenCalled();
  });

})