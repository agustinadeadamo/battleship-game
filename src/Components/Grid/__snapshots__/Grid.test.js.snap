// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Grid Component] Matches snapshot 1`] = `[Function]`;
